File: frontend/src/api/descriptive.ts
export interface StatRow {
  category: 'Quantiles' | 'Central Tendency' | 'Dispersion' | 'Shape' | 'Extremes'
  measure: string
  value: number | null
  consistencyCorr: number | null
  robust?: boolean
  advancedId?: string
  warning?: string
}

export interface DescriptiveSummary {
  n: number
  mean: number
  median: number
  std: number
  iqr: number
}

export interface HistogramData {
  binEdges: number[]
  counts: number[]
}

export interface DescriptiveResult {
  rows: StatRow[]
  summary: DescriptiveSummary
  histogram: HistogramData
  boxData: {
    min: number; q1: number; median: number; q3: number; max: number
    outliers: number[]
  }
}

export interface DescriptiveInput {
  sessionId: string
  column: string
  quantileProbs: number[]
  trimAlpha: number | null
  winsorLimits: [number, number] | null
  weightsCol: string | null
  filters: Record<string, string[]> | null
}

export async function computeDescriptiveStats(input: DescriptiveInput): Promise<DescriptiveResult> {
  const res = await fetch('/api/descriptive/compute', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-session-id': input.sessionId
    },
    body: JSON.stringify({
      session_id: input.sessionId,
      column: input.column,
      quantileProbs: input.quantileProbs,
      trimAlpha: input.trimAlpha,
      winsorLimits: input.winsorLimits,
      weightsCol: input.weightsCol,
      filters: input.filters
    })
  })

  if (!res.ok) {
    const text = await res.text()
    let errStr = text
    try {
      const errObj = JSON.parse(text)
      errStr = errObj.detail || text
    } catch {
      if (!text) errStr = `Backend proxy error: status ${res.status}`
    }
    throw new Error(errStr)
  }

  return res.json()
}

File: backend/api/schemas/descriptive.py
from pydantic import BaseModel
from typing import Dict, List, Optional, Tuple, Literal

class DescriptiveRequest(BaseModel):
    session_id: str
    column: str
    quantileProbs: List[float] = [0.25, 0.5, 0.75]
    trimAlpha: Optional[float] = None
    winsorLimits: Optional[Tuple[float, float]] = None
    weightsCol: Optional[str] = None
    filters: Optional[Dict[str, List[str]]] = None

class StatRow(BaseModel):
    category: Literal['Quantiles', 'Central Tendency', 'Dispersion', 'Shape', 'Extremes']
    measure: str
    value: float | None
    consistencyCorr: float | None
    robust: bool = False
    advancedId: Optional[str] = None
    warning: Optional[str] = None

class DescriptiveSummary(BaseModel):
    n: int
    mean: float
    median: float
    std: float
    iqr: float

class HistogramData(BaseModel):
    binEdges: List[float]
    counts: List[int]

class BoxData(BaseModel):
    min: float
    q1: float
    median: float
    q3: float
    max: float
    outliers: List[float]

class DescriptiveResponse(BaseModel):
    rows: List[StatRow]
    summary: DescriptiveSummary
    histogram: HistogramData
    boxData: BoxData
